dandiapi/api/zarr_checksums.py

```diff
@@ -0,0 +1,278 @@
+        if existing_zarr_checksum:
+            self._checksums = existing_zarr_checksum.checksums
+        else:
+            self._checksums = []
+        return self
+
+    def __exit__(self, exc_type, *exc):
+        # If there was an exception, do not write anything
+        if exc_type:
+            return None  # this means throw the exception as normal
+        if len(self.checksum_listing.checksums) > 0:
+            self.write_checksum_file(self.checksum_listing)
+        else:
+            # If there are no checksums to write, simply delete the checksum file.
+            self.delete_checksum_file()
+
+    @property
+    def checksum_file_path(self):
+        """Generate the path of the checksum file to update."""
+        return f'{settings.DANDI_DANDISETS_BUCKET_PREFIX}{settings.DANDI_ZARR_CHECKSUM_PREFIX_NAME}/{self.zarr_archive.zarr_id}/{self.zarr_directory_path}.checksum'  # noqa: E501
+
+    @property
+    def checksum_listing(self) -> ZarrChecksumListing:
+        """Get the current state of the updater."""
+        if self._checksums is None:
+            raise ValueError('This method is only valid when used by a context manager')
+        return self._serializer.generate_listing(self._checksums)
+
+    def read_checksum_file(self) -> ZarrChecksumListing | None:
+        """Load a checksum listing from the checksum file."""
+        from dandiapi.api.models import ZarrUploadFile
+
+        storage = ZarrUploadFile.blob.field.storage
+        checksum_path = self.checksum_file_path
+        if storage.exists(checksum_path):
+            with storage.open(checksum_path) as f:
+                x = f.read()
+                return self._serializer.deserialize(x)
+        else:
+            return None
+
+    def write_checksum_file(self, zarr_checksum: ZarrChecksumListing):
+        """Write a checksum listing to the checksum file."""
+        from dandiapi.api.models import ZarrUploadFile
+
+        storage = ZarrUploadFile.blob.field.storage
+        content_file = ContentFile(self._serializer.serialize(zarr_checksum).encode('utf-8'))
+        # save() will never overwrite an existing file, it simply appends some garbage to ensure
+        # uniqueness. _save() is an internal storage API that will overwite existing files.
+        storage._save(self.checksum_file_path, content_file)
+
+    def delete_checksum_file(self):
+        """Delete the checksum file."""
+        from dandiapi.api.models import ZarrUploadFile
+
+        storage = ZarrUploadFile.blob.field.storage
+        storage.delete(self.checksum_file_path)
+
+    def add_checksums(self, checksums: List[ZarrChecksum]):
+        """Add a list of checksums to the listing."""
+        if self._checksums is None:
+            raise ValueError('This method is only valid when used by a context manager')
+        for new_checksum in checksums:
+            try:
+                self._checksums[self._checksums.index(new_checksum)] = new_checksum
```

Comment:
O(n)

Reply:
also what happens if two files within zarr have the same checksum?

Reply:
`self._checksums` is a list, so probably nothing. That's definitely worth testing though, I will add some additional tests.

Reply:
Turns out I already tested that :+1: